File: visual-studio/source/GitWebLinks/Services/LinkHandlerProvider.cs
#nullable enable

using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace GitWebLinks;

public class LinkHandlerProvider {

    private readonly IReadOnlyCollection<ILinkHandler> _handlers;
    private readonly ILogger _logger;


    public LinkHandlerProvider(ISettings settings, Git git, ILogger logger) {
        _logger = logger;

        _handlers = DefinitionProvider
            .GetDefinitions()
            .OrderBy((x) => x.Name)
            .Select((x) => new LinkHandler(x, settings, git))
            .ToList();
    }


    public async Task<ILinkHandler?> SelectAsync(Repository repository) {
        if (repository.Remote is null) {
            return null;
        }

        await _logger.LogAsync($"Finding a handler for repository {repository.Remote}.");

        foreach (ILinkHandler handler in _handlers) {
            await _logger.LogAsync($"Testing '{handler.Name}");

            if (await handler.IsMatchAsync(repository.Remote.Url)) {
                await _logger.LogAsync($"Handler '{handler.Name}' is a match.");
                return handler;
            }
        }

        await _logger.LogAsync("No handler found.");
        return null;
    }


    public async Task<IReadOnlyCollection<UrlInfo>> GetUrlInfoAsync(string url) {
        IReadOnlyCollection<UrlInfo> output;


        await _logger.LogAsync($"Finding file info for URL '{url}'.");
        output = await InternalGetUrlInfoAsync(url, true);

        if (output.Count == 0) {
            await _logger.LogAsync("No strict matches found. Trying again with loose matching.");
            output = await InternalGetUrlInfoAsync(url, false);
        }

        return output;
    }


    private async Task<IReadOnlyCollection<UrlInfo>> InternalGetUrlInfoAsync(string url, bool strict) {
        List<UrlInfo> output;


        output = new List<UrlInfo>();

        foreach (ILinkHandler handler in _handlers) {
            UrlInfo? info;


            info = await handler.GetUrlInfoAsync(url, strict);

            if (info is not null) {
                await _logger.LogAsync($"The handler '{handler.Name}' mapped the file to '{info}'.");
                output.Add(info);
            }
        }

        return output;
    }

}

File: vscode/src/link-handler-provider.ts
import type { Git } from './git';
import type { RepositoryWithRemote, UrlInfo } from './types';

import { LinkHandler } from './link-handler';
import { log } from './log';
import { load } from './schema';

/**
 * Provides access to the link handlers.
 */
export class LinkHandlerProvider {
    private readonly handlers: readonly LinkHandler[];

    /**
     * @constructor
     * @param git The Git service.
     */
    public constructor(git: Git) {
        this.handlers = load()
            .sort((x, y) => x.name.localeCompare(y.name))
            .map((definition) => new LinkHandler(definition, git));
    }

    /**
     * Selects the link handler to generate links for the given repository.
     *
     * @param repository The repository to select the handler for.
     * @returns The handler to use and the remote URL that was matched, or `undefined` if the repository is not supported.
     */
    public select(repository: RepositoryWithRemote): SelectedLinkHandler | undefined {
        log('Finding a handler for repository %O.', repository.remote);

        for (let handler of this.handlers) {
            log("Testing '%s'.", handler.name);

            for (let url of repository.remote.urls) {
                if (handler.handlesRemoteUrl(url)) {
                    log("Handler '%s' is a match.", handler.name);
                    return { handler, remoteUrl: url };
                }
            }
        }

        log('No handler found.');
        return undefined;
    }

    /**
     * Gets the URL info from the handlers for the given URL.
     *
     * @param url The URL to get the info for.
     * @returns The URL info that was found.
     */
    public getUrlInfo(url: string): UrlInfo[] {
        let output: UrlInfo[];

        log("Finding file info for URL '%s'.", url);
        output = this.internalGetUrlInfo(url, true);

        if (output.length === 0) {
            log('No strict matches found. Trying again with loose matching.');
            output = this.internalGetUrlInfo(url, false);
        }

        return output;
    }

    /**
     * Gets the URL info from the handlers for the given URL using the specified mode.
     *
     * @param url The URL to get the info fo.
     * @param strict Whether to use strict matching.
     * @returns The URL info.
     */
    private internalGetUrlInfo(url: string, strict: boolean): UrlInfo[] {
        let output: UrlInfo[];

        output = [];

        for (let handler of this.handlers) {
            let info: UrlInfo | undefined;

            info = handler.getUrlInfo(url, strict);

            if (info) {
                log("The handler '%s' mapped the file to '%O'.", handler.name, info);
                output.push(info);
            }
        }

        return output;
    }
}

export interface SelectedLinkHandler {
    /**
     * The selected handler.
     */
    readonly handler: LinkHandler;

    /**
     * The remote URL that was used to select the handler.
     */
    readonly remoteUrl: string;
}
